### PR TITLE
major improvements to nanovdb

### PR DESCRIPTION
**Pull request by kmuseth**
Signed-off-by: Ken <ken.museth@gmail.com>

New features:

- Added nanovdb::IndexGrid that allows for arbitrary voxel values and even multiple channels to be associated with NanoVDB grids. They are more flexible and memory efficient than regular grids at a small cost of random-access performance. IndexGrids are constructed with the IndexGridBuilder and accessed with the new ChannelAccessor.
- Added iterators to tree node classes for visiting inactive, active, or all values and child nodes
- NanoVDB.h now includes standalone implementations of reading and writing uncompressed nanovdb grids buffers.
- Added Stats::getExtrema, which computes the min/max values of all voxels in a NanoVDB grid that intersects with a user-defined bounding-box.

Improvements:

- Added a new much improved nanovdb::NodeManager, that is both faster and more memory efficient, and that works on both the CPU and GPU. Note, it uses an handle for allocation (just like nanovdb::Grid) and replaces the old nanovdb::LeafManager which is now deprecated.
- NanoToOpenVDB now works with grids templated on Fp4, Fp8, Fp16, FpN, bool and ValueMask
- Numerous new member methods and utility functions were added, e.g. Mask::countOn(int) which is essential to the new IndexGrid
- PNanoVDB.h (a C99 port of NanoVDB.h) has been updated
- Rename RootNode::findTile to RootNode::probeTile and made it public
- Added RootNode::probeChild and InternalNode::probeChild
- Made Mask::getWord return both an const and non-const reference
- Improved unit-tests to use explicit 32B alignment

Bug fixes

- Fixed a bug in nanovdb::HostBuffer that could produce crashes due to misaligned CPU memory allocations.
- Fixed bug related to uninitialized memory in nanovdb::Grid which could confuse CRC32 checksum validation.
- Fixed bugs related to the use of intrinsic functions for fast bit-counting

### CODE REVIEW

**Comment by kmuseth** [comment]
The OSX builds consistently fails and I have a strong suspicion it has to do with the failure to create a directory named "data" before the unit tests (testNanoVDB and testOpenVDB) are run. Who configured the OSX CI ? Does anyone know how to check/fix this?

**Comment by Idclip** [comment]
> The OSX builds consistently fails and I have a strong suspicion it has to do with the failure to create a directory named "data" before the unit tests (testNanoVDB and testOpenVDB) are run. Who configured the OSX CI ? Does anyone know how to check/fix this?

You can see the Mac CI configuration in `.github/workflows/nanovdb.yml`. Look for the `macos-nanovdb` job. It's extremely basic and it's passing on master. As far as I can tell you didn't change anything with this 'data' directory - why do you think this is the cause?

Ideally you should be responsible for the nanovdb CI - please take a moment to try and familirise yourself with it and have a go at debuging

**Comment by kmuseth** [comment]
> 

fixed

File: nanovdb/nanovdb/unittest/pnanovdb_validate_strides.h [hunk 1]
```diff
@@ -83,26 +83,39 @@ static void compute_node_strides(
     pnanovdb_uint32_t minmaxStride = pnanovdb_grid_type_minmax_strides_bits[grid_type] / 8u;
     pnanovdb_uint32_t minmaxAlign = pnanovdb_grid_type_minmax_aligns_bits[grid_type] / 8u;
     pnanovdb_uint32_t statStride = pnanovdb_grid_type_stat_strides_bits[grid_type] / 8u;
-    if (nodeLevel == 0u && pnanovdb_grid_type_leaf_type[grid_type] == PNANOVDB_LEAF_TYPE_LITE)
+    pnanovdb_uint32_t postStatStride = 0u;
```
Comment (apradhana) [inline nanovdb/nanovdb/unittest/pnanovdb_validate_strides.h:86]: Looks OK to me.

File: nanovdb/nanovdb/util/GridBuilder.h [hunk 2]
```diff
@@ -541,21 +542,8 @@ GridHandle<BufferT> GridBuilder<ValueT, BuildT, StatsT>::
...
-    };
-    const double invMat[4][4] = {
-        {1 / dx, 0.0, 0.0, 0.0}, // row 0
-        {0.0, 1 / dx, 0.0, 0.0}, // row 1
-        {0.0, 0.0, 1 / dx, 0.0}, // row 2
-        {-Tx, -Ty, -Tz, 1.0}, // row 3
-    };
-    map.set(mat, invMat, 1.0);
+    Map map; // affine map
+    map.set(dx, p0, 1.0);
```
Comment (apradhana) [inline nanovdb/nanovdb/util/GridBuilder.h:546]: So convenient now!

File: nanovdb/nanovdb/util/GridStats.h [hunk 3]
```diff
@@ -730,6 +730,117 @@ void gridStats(NanoGrid<BuildT>& grid, StatsMode mode)
...
+        auto it = b.begin();// iterates over all the child nodes or tiles that intersects bbox
+        for (const Coord& ijk = *it; it; ++it) {
+            mask.setOn(ijk[2] + (ijk[1] << NodeT::LOG2DIM) + (ijk[0] << 2*NodeT::LOG2DIM));
+        }
+    }
+    return mask;
+}
+}// end of unnamed namespace
+
+/// @brief return the extrema of all the values in a grid that
```
Comment (apradhana) [inline nanovdb/nanovdb/util/GridStats.h:764]: This chunk of code has been reviewed in another PR.

File: nanovdb/nanovdb/util/Stencils.h [hunk 4]
```diff
@@ -225,13 +225,13 @@ class BaseStencil
...
         uint8_t bits;
-        Mask() : bits(0u) {}
-        void set(int i) { bits |= (1 << i); }
-        bool test(int i) const { return bits & (1 << i); }
-        bool any() const  { return bits >  0u; }
-        bool all() const  { return bits == 255u; }
-        bool none() const { return bits == 0u; }
-        int count() const { return CountOn(bits); }
+        __hostdev__ Mask() : bits(0u) {}
+        __hostdev__ void set(int i) { bits |= (1 << i); }
```
Comment (apradhana) [inline nanovdb/nanovdb/util/Stencils.h:229]: Useful.

File: nanovdb/nanovdb/util/NodeManager.h [hunk 5]
```diff
@@ -8,393 +8,311 @@
...
-{
-}
+    /// @warning Note that the return pointer can be NULL if the template parameter does not match the specified grid!
+    template<typename BuildT, typename U = BufferT>
+    typename std::enable_if<BufferTraits<U>::hasDeviceDual, const NodeManager<BuildT>*>::type
+    deviceMgr() const { return this->template getDeviceMgr<BuildT>(); }
 
-template<typename GridT>
-NodeManager<GridT>::NodeManager(GridT &grid)
-    : mGrid(&grid)
```
Comment (apradhana) [inline nanovdb/nanovdb/util/NodeManager.h:141]: I see that we can no longer create a node manager this way: 
`nanovdb::NodeManager<nanovdb::FloatGrid> nodeMgr(*host_grid);`
Should we make a note in createNodeManager that this is the preferred way to construct a NodeManager now?
```
    auto node_mgr_handle = nanovdb::createNodeManager(*host_grid);
    auto *mgr = node_mgr_handle.mgr<float>();
```

**Review by apradhana — APPROVED** [review]

**Pull request merged**